Sema: Improve redeclaration error for synthesized inits

AnthonyLatsis opened the pull request:
Resolves [SR-8021](https://bugs.swift.org/browse/SR-8021).

@jrose-apple

AnthonyLatsis committed:
Improve redeclaration error for implicit inits

jrose-apple commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -914,8 +914,15 @@ static void checkRedeclaration(TypeChecker &tc, ValueDecl *current) {
-        tc.diagnose(current, diag::invalid_redecl, current->getFullName());
-        tc.diagnose(other, diag::invalid_redecl_prev, other->getFullName());
+        const auto *otherInit = dyn_cast<ConstructorDecl>(other);
+        if (otherInit && otherInit->isImplicit()) {
```
This isn't the only reason a constructor can be implicit; it might be inherited. Maybe check that it's not overriding anything?

AnthonyLatsis replied:
Good point.

jrose-apple commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -667,6 +667,9 @@ ERROR(reserved_member_name,none,
 ERROR(invalid_redecl,none,"invalid redeclaration of %0", (DeclName))
+ERROR(invalid_redecl_synth_init,none,
+      "invalid redeclaration of synthesized "
+      "%select{default|memberwise}1 %0", (DeclName, bool))
```
Do we actually use the term "default initializer" for a no-argument initializer? Unlike in C++, it's not the "default" if you don't initialize something. It might be fine just to leave out the qualifier in that case.

AnthonyLatsis replied:
OK. I tend to call "default" an synthesized no-argument initializer.

AnthonyLatsis replied:
[Heh](https://docs.swift.org/swift-book/LanguageGuide/Initialization.html#ID213) :)

jrose-apple replied:
I stand corrected, and it makes sense if it only refers to the synthesized one.

jrose-apple reviewed (commented):
Thanks for picking this up!

AnthonyLatsis committed:
Skip redecl errors for inits that conflict with an inherited init

This can only happen when the conflicting init is within an extension.
It is the override checker's responsibility to emit a sensible diagnostic here.

AnthonyLatsis commented:
So I've gone a bit further with the second commit disabling redeclarations for inherited inits when they conflict with the implicit one. I reason is that the redeclaration error here always intersects with the more productive "cannot override in extension", which in my opinion can easily confuse.

jrose-apple commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -810,6 +810,13 @@ static void checkRedeclaration(TypeChecker &tc, ValueDecl *current) {
+      // have already taken care of emitting a more productive diagnostic.
+      if (isa<ConstructorDecl>(other) && other->isImplicit() &&
+          other->getOverriddenDecl())
+        continue;
```
I'm concerned by this not requiring any changes to existing tests, and I don't think your new test covers it. What if the inherited initializer *is* exposed to Objective-C?

jrose-apple replied:
Also, I'm not sure exiting early is a good idea, because we still need to mark one of them invalid.

AnthonyLatsis replied:
Sema, Constraints, decl and Generics pass just fine, but it might affect something elsewhere. 

> and I don't think your new test covers it. What if the inherited initializer is exposed to Objective-C?

You're right, I'll add some more tests with `@objc` and `dynamic`. Right now though it boils down to `'init(arg:)' has already been overridden` with and without the change if I plug all the errors.

AnthonyLatsis commented:
I have no idea if there is a use case, but should we allow to override an `@objc dynamic` designated init with a convenience init in an *extension*? An analogy with a method is apparently possible.

jrose-apple commented:
Unfortunately I think that's legal, but not if the initializers have already been inherited. In this case, though, source compat is probably the deciding factor. What does Swift 5.0 do?

AnthonyLatsis commented:
> What does Swift 5.0 do?

When I try to override one? `'init' has already been overridden`. I added a test for that, by the way.

jrose-apple reviewed (approved):
Okay, this looks good, then. Thanks, @AnthonyLatsis!

jrose-apple commented:
@swift-ci Please smoke test

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
@swift-ci Please test source compatibility

AnthonyLatsis commented:
Thank you for guiding me through this!

Just in case there *was* a misunderstanding:
```swift
class A {
  @objc dynamic init(arg: Int) {}
}
class B: A {}
extension B {
  convenience override init(arg: Int) {} // 'init(arg:)' has already been overridden
}
```
I was pondering solutions to the "already overridden" diagnostic (which isn't great in this case), and then realized it might be legal to simply allow the override to follow, since the same can be done to a method. Hence the question earlier /.

jrose-apple commented:
Oops. Can you add `-enable-objc-interop -disable-objc-attr-requires-foundation-module` to the RUN line for this test, so that it'll work on Linux?

AnthonyLatsis committed:
Update basic_init.swift

jrose-apple commented:
@swift-ci Please smoke test
